File: src/main/java/com/ealanta/productapp/web/ExceptionTranslator.java
package com.ealanta.productapp.web;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;

import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@ControllerAdvice
public class ExceptionTranslator {

    @ExceptionHandler(MethodArgumentTypeMismatchException.class)
    @ResponseBody
    public ResponseEntity<String> handleMethodArgumentTypeMismatchException(MethodArgumentTypeMismatchException e) {
        Class<?> type = e.getRequiredType();
        e.getRequiredType();
        final String message;
        if(type.isEnum()){
            Stream<String> stream = Arrays.stream(type.getEnumConstants())
                                            .map(Object::toString);
            String withComma=stream.collect(Collectors.joining(", "));
            message = "The parameter " + e.getName() + " must have a value among : " + withComma;
        } else {
            message = "The parameter " + e.getName() + " must be of type " + type.getTypeName();
        }
        return ResponseEntity.unprocessableEntity().body(message);
    }
}

File: src/main/kotlin/com/ealanta/productapp/web/ExceptionTranslator.kt
package com.ealanta.productapp.web

import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException

@ControllerAdvice
class ExceptionTranslator {

    @ExceptionHandler(MethodArgumentTypeMismatchException::class)
    @ResponseBody
    fun handleMethodArgumentTypeMismatchException(e: MethodArgumentTypeMismatchException): ResponseEntity<String> {
        val type = e.requiredType
        e.requiredType
        val message = if (type!!.isEnum) {
            val withComma = type.enumConstants
                    .map { it.toString() }
                    .joinToString(separator= ", ")
            "The parameter ${e.name} must have a value among : $withComma"
        } else {
            "The parameter ${e.name} must be of type ${type.typeName}"
        }
        return ResponseEntity.unprocessableEntity().contentType(MediaType.TEXT_PLAIN).body(message)
    }
}